source/websocket.c

```diff
@@ -0,0 +1,908 @@
+
+                s_stop_writing(websocket, AWS_ERROR_HTTP_CONNECTION_CLOSED);
+                s_finish_shutdown(websocket);
+            } else {
+                AWS_LOGF_TRACE(
+                    AWS_LS_HTTP_WEBSOCKET,
+                    "id=%p: CLOSE frame queued, handler will finish shutdown once it's sent.",
+                    (void *)websocket);
+            }
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static void s_finish_shutdown(struct aws_websocket *websocket) {
+    assert(aws_channel_thread_is_callers_thread(websocket->channel_slot->channel));
+    assert(websocket->thread_data.is_writing_stopped);
+    assert(websocket->thread_data.is_shutting_down_and_waiting_for_close_frame_to_be_written);
+
+    AWS_LOGF_TRACE(AWS_LS_HTTP_WEBSOCKET, "id=%p: Finishing websocket handler shutdown.", (void *)websocket);
+
+    websocket->thread_data.is_shutting_down_and_waiting_for_close_frame_to_be_written = false;
+
+    /* Cancel all incomplete outgoing frames */
+    if (websocket->thread_data.current_outgoing_frame) {
+        s_destroy_outgoing_frame(
+            websocket, websocket->thread_data.current_outgoing_frame, AWS_ERROR_HTTP_CONNECTION_CLOSED);
+        websocket->thread_data.current_outgoing_frame = NULL;
+    }
+
+    /* BEGIN CRITICAL SECTION */
+    int err = aws_mutex_lock(&websocket->synced_data.lock);
+    assert(!err);
+    (void)err;
+
+    while (!aws_linked_list_empty(&websocket->synced_data.outgoing_frame_list)) {
+        /* Move frames from synced_data to thread_data, then cancel them together outside critical section */
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&websocket->synced_data.outgoing_frame_list);
+        aws_linked_list_push_back(&websocket->thread_data.outgoing_frame_list, node);
+    }
+
+    err = aws_mutex_unlock(&websocket->synced_data.lock);
+    assert(!err);
+    (void)err;
+    /* END CRITICAL SECTION */
+
+    while (!aws_linked_list_empty(&websocket->thread_data.outgoing_frame_list)) {
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&websocket->thread_data.outgoing_frame_list);
+        struct outgoing_frame *frame = AWS_CONTAINER_OF(node, struct outgoing_frame, node);
+        s_destroy_outgoing_frame(websocket, frame, AWS_ERROR_HTTP_CONNECTION_CLOSED);
+    }
+
+    /* TODO: cancel incomplete incoming_frame */
+
+    if (websocket->on_connection_shutdown) {
+        AWS_LOGF_TRACE(AWS_LS_HTTP_WEBSOCKET, "id=%p: Invoking user's shutdown callback.", (void *)websocket);
+        websocket->on_connection_shutdown(
+            websocket, websocket->thread_data.channel_shutdown_error_code, websocket->user_data);
+    }
+
+    aws_channel_slot_on_handler_shutdown_complete(
+        websocket->channel_slot,
+        AWS_CHANNEL_DIR_WRITE,
+        websocket->thread_data.channel_shutdown_error_code,
+        websocket->thread_data.channel_shutdown_free_scarce_resources_immediately);
+}
+
+static int s_handler_process_read_message(
```

Comment:
huh? how does this even work without this one implementated?

Reply:
It only works in unit tests so far.
Next PR will deal with incoming data